zerver/lib/streams.py

```diff
@@ -181,20 +182,49 @@ def get_default_values_for_stream_permission_group_settings(
     return group_setting_values
 
 
-def get_user_ids_with_metadata_access_via_permission_groups(stream: Stream) -> set[int]:
-    return set(
-        get_recursive_group_members_union_for_groups(
-            [stream.can_add_subscribers_group_id, stream.can_administer_channel_group_id]
-        )
+def get_users_dict_with_metadata_access_to_streams_via_permission_groups(
+    streams: list[Stream],
+    realm_id: int,
+) -> dict[int, set[int]]:
+    can_administer_group_ids = {stream.can_administer_channel_group_id for stream in streams}
+    can_add_subscriber_group_ids = {stream.can_add_subscribers_group_id for stream in streams}
+
+    all_permission_group_ids = list(can_administer_group_ids | can_add_subscriber_group_ids)
+
+    recursive_group_members = (
+        get_root_id_annotated_recursive_subgroups_for_groups(all_permission_group_ids, realm_id)
+        .filter(direct_members__is_active=True)
         .exclude(
             # allow_everyone_group=False is false for both
             # can_add_subscribers_group and
             # can_administer_channel_group, so guest users cannot
             # exercise these permission to get metadata access.
-            role=UserProfile.ROLE_GUEST
+            direct_members__role=UserProfile.ROLE_GUEST
+        )
+        .values_list("root_id", "direct_members")  # type: ignore[misc]  # root_id is an annotated field.
```

Comment:
Hmm, does this end up excluding guests, or groups that have a guest among their direct members? I think I'd expect the latter, which would not be correct...

Reply:
Yes, groups that have a guest is being excluded. I think then code in `mention.py` is also incorrect, will verify it once.

Correct method then is to fetch members in a separate query other than a query to fetch recursive subgroups. But this will lead to two queries even when computing users for one stream, where we can do that in 1 query. So, one optimization we can do is to just have a separate block at top when there is only one stream.

Reply:
Worth checking. I think we did add a test for this sort of corner case, but we could have done it wrong.

Reply:
Checked the mention code. It is correct, since group is excluded only if all the members are deactivated which is fine.